### PR TITLE
server-side-website with additional S3 Origin behaviour

### PR DESCRIPTION
**Pull request by aran112000**
This adds support to create additional CloudFront Origins & Behaviours to you can map additional paths to alternative, existing S3 buckets.

Includes updated documentation to show how this can be used.

### CODE REVIEW

File: docs/server-side-website.md [hunk 1]
```diff
@@ -343,6 +343,19 @@ Applications are of course free to catch errors and display custom error pages.
 
 All headers are forwarded to your Lambda backend. If you used the `forwardedHeaders` option in the past, you can safely remove it.
 
+## Routing paths to other S3 Origins
```
Comment (mnapoli) [inline docs/server-side-website.md:346]: This should be a `###`. Also a suggestion to make the title less low-level:

```suggestion
### Routing paths to other S3 buckets
```

File: docs/server-side-website.md [hunk 2]
```diff
@@ -343,6 +343,19 @@ Applications are of course free to catch errors and display custom error pages.
...
+
+If you need to route any paths to alternative S3 Origins, you can declare the S3 origin as a source, like you would with a local directly, just prefixing the S3 bucket name with `s3://`.
```
Comment (mnapoli) [inline docs/server-side-website.md:348]:
```suggestion
You can route any paths to alternative S3 buckets like you would with a local directory. To do that, prefix the S3 bucket name with `s3://`.
```

Can we also add a sentence on how the S3 bucket should be configured? I'm guessing it needs to be either public, or have a origin policy set up?

File: docs/server-side-website.md [hunk 3]
```diff
@@ -343,6 +343,19 @@ Applications are of course free to catch errors and display custom error pages.
...
+
+```yaml
+constructs:
+    website:
+        # ...
+        assets:
+            '/assets/*': dist/
+            '/uploads/*: s://another-bucket-name
```
Comment (mnapoli) [inline docs/server-side-website.md:356]:
```suggestion
            '/uploads/*: s3://another-bucket-name
```

File: src/constructs/aws/ServerSideWebsite.ts [hunk 4]
```diff
@@ -240,6 +240,10 @@ export class ServerSideWebsite extends AwsConstruct {
 
         let invalidate = false;
         for (const [pattern, filePath] of Object.entries(this.getAssetPatterns())) {
+            if (filePath.substring(0, 5) === 's3://') {
```
Comment (mnapoli) [inline src/constructs/aws/ServerSideWebsite.ts:243]:
```suggestion
            // Ignore external buckets
            if (filePath.substring(0, 5) === 's3://') {
```

File: src/constructs/aws/ServerSideWebsite.ts [hunk 5]
```diff
@@ -372,7 +384,7 @@ export class ServerSideWebsite extends AwsConstruct {
     private createRequestFunction(): cloudfront.Function {
         let additionalCode = "";
 
-        if (this.configuration.redirectToMainDomain === true) {
+        if (this.configuration.redirectToMainDomain === true && this.domains?.length > 1) {
```
Comment (mnapoli) [inline src/constructs/aws/ServerSideWebsite.ts:387]: This duplicates #321, correct?

I think this should be reverted from this PR to avoid conflicts and keep things on topic.

**Review by mnapoli — COMMENTED** [review]
Awesome, that looks like a great idea that will benefit many! (the "/uploads/" example is very common)

I added some suggestions inline. Could you also add some tests? This is a large feature so we want to make sure it is covered.

File: docs/server-side-website.md [hunk 6]
```diff
@@ -343,6 +343,45 @@ Applications are of course free to catch errors and display custom error pages.
...
+
+You can route any paths to alternative S3 buckets like you would with a local directory. To do that, prefix the S3 bucket name with `s3://`.
+
+```yaml
+constructs:
+    website:
+        # ...
+        assets:
+            '/assets/*': dist/
+            '/uploads/*: s3://another-bucket-name
```
Comment (mnapoli) [inline docs/server-side-website.md:356]: Just noticed a missing quote thanks to the syntax highlighting:

```suggestion
            '/uploads/*': s3://another-bucket-name
```

**Comment by mnapoli** [comment]
Hi, just bumping here regarding my last comment: I think the only thing missing here for a merge are tests? Especially since you discovered plenty of edge cases, maintaining this feature sounds like it will be quite some effort, so the tests will help minimize the effort 😅

**Comment by aran112000** [comment]
> Hi, just bumping here regarding my last comment: I think the only thing missing here for a merge are tests? Especially since you discovered plenty of edge cases, maintaining this feature sounds like it will be quite some effort, so the tests will help minimize the effort 😅

Sure thing, @robchett just added these now :)

**Comment by evo-chris** [comment]
@mnapoli Just wondering if there is a plan in mind for the next release and will it include this PR? If there's anything we can help with, let us know. Thanks.

**Comment by aran112000** [comment]
Hey @fredericbarthelet, @mnapoli 

Any chance this could be reviewed / merged if you're happy with the change & tests added here please?

**Comment by kevincerro** [comment]
Hope that this will be finally merged

**Comment by vikivyas** [comment]
Aran Reeks (@aran112000)  while It is not being approved, how can we use this? 

will uploaded files to the bucket be removed on the next deployment?

**Comment by aran112000** [comment]
> Aran Reeks (@aran112000) while It is not being approved, how can we use this?
> 
> will uploaded files to the bucket be removed on the next deployment?

Feel free to use my fork of Lift, we use this on production on a number of sites successfully. It doesn't clear any assets from other buckets on deployment. They're just added as origins to your Cloudfront distributions.

@mnapoli - is it possible to complete the review and get this merged, please?

**Comment by hschimpf** [comment]
This is just what I need! 🚀 

I was adding an additional Origin+Behavior to the distribution manually.

Any updates on this? We'll really love this to be merged.

**Comment by vikivyas** [comment]
> > Aran Reeks (@aran112000) while It is not being approved, how can we use this?
> > will uploaded files to the bucket be removed on the next deployment?
> 
> Feel free to use my fork of Lift, we use this on production on a number of sites successfully. It doesn't clear any assets from other buckets on deployment. They're just added as origins to your Cloudfront distributions.
> 
> @mnapoli - is it possible to complete the review and get this merged, please?

can you share any of your serverless.yml as an example? I am looking for best example for my laravel application.

**Comment by hschimpf** [comment]
> can you share any of your serverless.yml as an example? I am looking for best example for my laravel application.

Hey @vikivyas, here is mine from a currently live production application: [serverless.yml](https://gist.github.com/hschimpf/b6bc251f36f14d25e1f153b37f6b7139)

I have two buckets, one is created for the frontend assets, and another is private for customer's uploads. There are some uploads that should be able to be publicly accessible _(like profile pics, personalized banners, etc)_, those files are uploaded to a folder called `c/` on the private bucket, and that folder is specified as another Origin+Behavior on the `server-side-website` construct:

```yml
constructs:
  assets:
    type: server-side-website
    domain: ${param:domain}
    certificate: ${param:certificateArn}
    assets:
      '/build/*': public/build
      '/robots.txt': public/robots.txt
      '/c/*': 's3://${param:UUID}-${aws:region}-storage'
    extensions:
      bucket:
        Properties:
          BucketName: '${param:UUID}-${aws:region}-assets'

  storage:
    type: storage
    extensions:
      bucket:
        DeletionPolicy: Retain
        Properties:
          BucketName: '${param:UUID}-${aws:region}-storage'
          AccessControl: Private
```

Let me know if it helps you
